File: kotlin/proximity/src/main/kotlin/com/hopae/eudi/wallet/proximity/Ndef.kt
package com.hopae.eudi.wallet.proximity

import java.io.ByteArrayOutputStream

/** One NDEF record (NFC Forum NDEF spec). [tnf] is the 3-bit Type Name Format; type/id/payload are raw bytes. */
class NdefRecord(val tnf: Int, val type: ByteArray, val id: ByteArray = ByteArray(0), val payload: ByteArray)

/**
 * Minimal NDEF message encoder/decoder — enough for ISO/IEC 18013-5 NFC static handover (a Handover Select
 * record, the DeviceEngagement record, and a BLE carrier-configuration record). No chunking.
 */
object Ndef {
    const val TNF_WELL_KNOWN = 0x01
    const val TNF_MIME_MEDIA = 0x02
    const val TNF_EXTERNAL = 0x04

    fun encodeMessage(records: List<NdefRecord>): ByteArray {
        val out = ByteArrayOutputStream()
        records.forEachIndexed { i, r -> out.write(encodeRecord(r, first = i == 0, last = i == records.size - 1)) }
        return out.toByteArray()
    }

    fun decodeMessage(bytes: ByteArray): List<NdefRecord> {
        val records = mutableListOf<NdefRecord>()
        var i = 0
        while (i < bytes.size) {
            val flags = bytes[i++].toInt() and 0xFF
            val tnf = flags and 0x07
            val sr = flags and 0x10 != 0
            val il = flags and 0x08 != 0
            val typeLen = bytes[i++].toInt() and 0xFF
            val payloadLen = if (sr) {
                bytes[i++].toInt() and 0xFF
            } else {
                val l = ((bytes[i].toInt() and 0xFF) shl 24) or ((bytes[i + 1].toInt() and 0xFF) shl 16) or
                    ((bytes[i + 2].toInt() and 0xFF) shl 8) or (bytes[i + 3].toInt() and 0xFF)
                i += 4; l
            }
            val idLen = if (il) bytes[i++].toInt() and 0xFF else 0
            val type = bytes.copyOfRange(i, i + typeLen); i += typeLen
            val id = bytes.copyOfRange(i, i + idLen); i += idLen
            val payload = bytes.copyOfRange(i, i + payloadLen); i += payloadLen
            records.add(NdefRecord(tnf, type, id, payload))
            if (flags and 0x40 != 0) break // ME (last record)
        }
        return records
    }

    private fun encodeRecord(r: NdefRecord, first: Boolean, last: Boolean): ByteArray {
        val sr = r.payload.size < 256
        val il = r.id.isNotEmpty()
        var flags = r.tnf
        if (first) flags = flags or 0x80 // MB
        if (last) flags = flags or 0x40 // ME
        if (sr) flags = flags or 0x10 // SR
        if (il) flags = flags or 0x08 // IL
        val out = ByteArrayOutputStream()
        out.write(flags)
        out.write(r.type.size)
        if (sr) {
            out.write(r.payload.size)
        } else {
            val n = r.payload.size
            out.write((n ushr 24) and 0xFF); out.write((n ushr 16) and 0xFF); out.write((n ushr 8) and 0xFF); out.write(n and 0xFF)
        }
        if (il) out.write(r.id.size)
        out.write(r.type)
        if (il) out.write(r.id)
        out.write(r.payload)
        return out.toByteArray()
    }
}

File: swift/Sources/Proximity/Ndef.swift
/// One NDEF record (NFC Forum NDEF spec). `tnf` is the 3-bit Type Name Format; type/id/payload are raw bytes.
public struct NdefRecord {
    public let tnf: Int
    public let type: [UInt8]
    public let id: [UInt8]
    public let payload: [UInt8]
    public init(tnf: Int, type: [UInt8], id: [UInt8] = [], payload: [UInt8]) {
        self.tnf = tnf; self.type = type; self.id = id; self.payload = payload
    }
}

/// Minimal NDEF message encoder/decoder — enough for ISO/IEC 18013-5 NFC static handover. No chunking.
public enum Ndef {
    public static let tnfWellKnown = 0x01
    public static let tnfMimeMedia = 0x02
    public static let tnfExternal = 0x04

    public static func encodeMessage(_ records: [NdefRecord]) -> [UInt8] {
        var out: [UInt8] = []
        for (i, r) in records.enumerated() { out += encodeRecord(r, first: i == 0, last: i == records.count - 1) }
        return out
    }

    public static func decodeMessage(_ bytes: [UInt8]) -> [NdefRecord] {
        var records: [NdefRecord] = []
        var i = 0
        while i < bytes.count {
            let flags = Int(bytes[i]); i += 1
            let tnf = flags & 0x07
            let sr = flags & 0x10 != 0
            let il = flags & 0x08 != 0
            let typeLen = Int(bytes[i]); i += 1
            let payloadLen: Int
            if sr {
                payloadLen = Int(bytes[i]); i += 1
            } else {
                payloadLen = (Int(bytes[i]) << 24) | (Int(bytes[i + 1]) << 16) | (Int(bytes[i + 2]) << 8) | Int(bytes[i + 3]); i += 4
            }
            let idLen = il ? Int(bytes[i]) : 0
            if il { i += 1 }
            let type = Array(bytes[i..<i + typeLen]); i += typeLen
            let id = Array(bytes[i..<i + idLen]); i += idLen
            let payload = Array(bytes[i..<i + payloadLen]); i += payloadLen
            records.append(NdefRecord(tnf: tnf, type: type, id: id, payload: payload))
            if flags & 0x40 != 0 { break } // ME (last record)
        }
        return records
    }

    private static func encodeRecord(_ r: NdefRecord, first: Bool, last: Bool) -> [UInt8] {
        let sr = r.payload.count < 256
        let il = !r.id.isEmpty
        var flags = r.tnf
        if first { flags |= 0x80 } // MB
        if last { flags |= 0x40 } // ME
        if sr { flags |= 0x10 } // SR
        if il { flags |= 0x08 } // IL
        var out: [UInt8] = [UInt8(flags), UInt8(r.type.count)]
        if sr {
            out.append(UInt8(r.payload.count))
        } else {
            let n = r.payload.count
            out += [UInt8((n >> 24) & 0xFF), UInt8((n >> 16) & 0xFF), UInt8((n >> 8) & 0xFF), UInt8(n & 0xFF)]
        }
        if il { out.append(UInt8(r.id.count)) }
        out += r.type
        if il { out += r.id }
        out += r.payload
        return out
    }
}
